components/nimbus/src/tests/test_behavior.rs

```diff
@@ -0,0 +1,657 @@
+
+        assert!(matches!(
+            counter
+                .intervals
+                .get(&Interval::Months)
+                .unwrap()
+                .data
+                .buckets[0],
+            1
+        ));
+        assert!(matches!(
+            counter.intervals.get(&Interval::Days).unwrap().data.buckets[0],
+            1
+        ));
+        Ok(())
+    }
+
+    #[test]
+    fn test_advance_do_not_advance() -> Result<()> {
+        let mut counter = MultiIntervalCounter::new(vec![
+            SingleIntervalCounter::new(IntervalConfig::new(12, Interval::Months)),
+            SingleIntervalCounter::new(IntervalConfig::new(28, Interval::Days)),
+        ]);
+        let date = Utc::now();
+        counter.maybe_advance(date).ok();
+
+        assert!(matches!(
+            counter
+                .intervals
+                .get(&Interval::Months)
+                .unwrap()
+                .data
+                .buckets
+                .len(),
+            1
+        ));
+        assert!(matches!(
+            counter
+                .intervals
+                .get(&Interval::Days)
+                .unwrap()
+                .data
+                .buckets
+                .len(),
+            1
+        ));
+        Ok(())
+    }
+
+    #[test]
+    fn test_advance_advance_some() -> Result<()> {
+        let mut counter = MultiIntervalCounter::new(vec![
+            SingleIntervalCounter::new(IntervalConfig::new(12, Interval::Months)),
+            SingleIntervalCounter::new(IntervalConfig::new(28, Interval::Days)),
+        ]);
+        let date = Utc::now() + Duration::days(1);
+        counter.maybe_advance(date).ok();
+
+        assert!(matches!(
+            counter
+                .intervals
+                .get(&Interval::Months)
+                .unwrap()
+                .data
+                .buckets
+                .len(),
+            1
+        ));
+        assert!(matches!(
+            counter
+                .intervals
+                .get(&Interval::Days)
+                .unwrap()
+                .data
+                .buckets
+                .len(),
+            2
+        ));
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod event_store_tests {
+    use chrono::Duration;
+
+    use super::*;
+
+    #[test]
+    fn record_event_should_function() -> Result<()> {
+        let counter1 = MultiIntervalCounter::new(vec![
+            SingleIntervalCounter::new(IntervalConfig::new(12, Interval::Months)),
+            SingleIntervalCounter::new(IntervalConfig::new(28, Interval::Days)),
+        ]);
+
+        let counter2 = MultiIntervalCounter::new(vec![
+            SingleIntervalCounter::new(IntervalConfig::new(12, Interval::Months)),
+            SingleIntervalCounter::new(IntervalConfig::new(28, Interval::Days)),
+        ]);
+
+        let mut store = EventStore::from(vec![
+            ("event-1".to_string(), counter1),
+            ("event-2".to_string(), counter2),
+        ]);
```

Comment:
Side question: how do new events get multi-counters, apart from at construction?

Reply:
I don't know that we've thought about that yet, but you're right we will need something in place to handle that as we perform releases and start to add new events.

Reply:
Ideally, I see this happening automatically in `record_event`, if it can't find an existing counter it would create one.

Reply:
Do we want to include that as a part of this PR or shall I include it in the next PR I'm opening?